httpx/concurrency/asyncio.py

```diff
@@ -247,28 +219,84 @@ def run(
     def get_semaphore(self, limits: PoolLimits) -> BasePoolSemaphore:
         return PoolSemaphore(limits)
 
-    def background_manager(
-        self, coroutine: typing.Callable, args: typing.Any
-    ) -> "BackgroundManager":
-        return BackgroundManager(coroutine, args)
+    def create_event(self) -> BaseEvent:
+        return typing.cast(BaseEvent, asyncio.Event())
+
+    def background_manager(self) -> "BackgroundManager":
+        return BackgroundManager()
+
+    def body_iterator(self) -> "BodyIterator":
+        return BodyIterator()
 
 
 class BackgroundManager(BaseBackgroundManager):
-    def __init__(self, coroutine: typing.Callable, args: typing.Any) -> None:
-        self.coroutine = coroutine
-        self.args = args
+    def __init__(self) -> None:
+        self.tasks: typing.Set[asyncio.Task] = set()
 
-    async def __aenter__(self) -> "BackgroundManager":
+    def start_soon(self, coroutine: typing.Callable, *args: typing.Any) -> None:
         loop = asyncio.get_event_loop()
-        self.task = loop.create_task(self.coroutine(*self.args))
-        return self
+        self.tasks.add(loop.create_task(coroutine(*args)))
+
+    def wait_first_completed(self) -> "WaitFirstCompleted":
+        return WaitFirstCompleted(self)
 
     async def __aexit__(
         self,
         exc_type: typing.Type[BaseException] = None,
         exc_value: BaseException = None,
         traceback: TracebackType = None,
     ) -> None:
-        await self.task
-        if exc_type is None:
-            self.task.result()
+        done, pending = await asyncio.wait(self.tasks, timeout=0)
+
+        for task in pending:
+            try:
+                # Give it a chance to finish soon
+                await asyncio.wait_for(task, timeout=0.01)
```

Comment:
If it actually does finish should we move it to `done`?

Reply:
Yup. Which made me realize we can simplify this by passing the small `timeout` to the first `.wait()` call instead. :+1: